File: client/java/client/src/main/java/dev/angzarr/client/router/CommandHandlerRouter.java
package dev.angzarr.client.router;

import com.google.protobuf.Any;
import com.google.protobuf.InvalidProtocolBufferException;
import dev.angzarr.*;
import dev.angzarr.client.Helpers;
import dev.angzarr.client.compensation.RejectionHandlerResponse;

import java.util.AbstractMap;
import java.util.List;
import java.util.Map;

/**
 * Router for command handler components (commands -> events, single domain).
 *
 * <p>Domain is set at construction time. No additional domain registration
 * is possible, enforcing single-domain constraint.
 *
 * <p>Example:
 * <pre>{@code
 * CommandHandlerRouter<PlayerState> router = new CommandHandlerRouter<>(
 *     "player",           // router name
 *     "player",           // domain
 *     new PlayerHandler() // handler
 * );
 *
 * // Get subscriptions for registration
 * List<Map.Entry<String, List<String>>> subs = router.subscriptions();
 *
 * // Dispatch a command
 * BusinessResponse response = router.dispatch(contextualCommand);
 * }</pre>
 *
 * @param <S> The state type for this command handler
 */
public class CommandHandlerRouter<S> {

    private final String name;
    private final String domain;
    private final CommandHandlerDomainHandler<S> handler;

    /**
     * Create a new command handler router.
     *
     * @param name The router name
     * @param domain The domain this command handler handles
     * @param handler The domain handler
     */
    public CommandHandlerRouter(String name, String domain, CommandHandlerDomainHandler<S> handler) {
        this.name = name;
        this.domain = domain;
        this.handler = handler;
    }

    /**
     * Get the router name.
     */
    public String getName() {
        return name;
    }

    /**
     * Get the domain.
     */
    public String getDomain() {
        return domain;
    }

    /**
     * Get command types from the handler.
     */
    public List<String> getCommandTypes() {
        return handler.commandTypes();
    }

    /**
     * Get subscriptions for this command handler.
     *
     * @return List of (domain, command types) pairs
     */
    public List<Map.Entry<String, List<String>>> subscriptions() {
        return List.of(new AbstractMap.SimpleEntry<>(domain, handler.commandTypes()));
    }

    /**
     * Rebuild state from events using the handler's state router.
     *
     * @param events The event book containing prior events
     * @return The rebuilt state
     */
    public S rebuildState(EventBook events) {
        return handler.rebuild(events);
    }

    /**
     * Dispatch a contextual command to the handler.
     *
     * @param cmd The contextual command containing command and prior events
     * @return The business response containing resulting events or rejection
     * @throws RouterException if dispatch fails
     */
    public BusinessResponse dispatch(ContextualCommand cmd) throws RouterException {
        // Validate command structure
        CommandBook commandBook = cmd.getCommand();
        if (commandBook == null || commandBook.getPagesList().isEmpty()) {
            throw new RouterException("Missing command book or pages");
        }

        CommandPage commandPage = commandBook.getPages(0);
        if (!commandPage.hasCommand()) {
            throw new RouterException("Missing command payload");
        }

        Any commandAny = commandPage.getCommand();
        EventBook eventBook = cmd.hasEvents() ? cmd.getEvents() : EventBook.getDefaultInstance();

        // Rebuild state
        S state = handler.rebuild(eventBook);
        int seq = Helpers.nextSequence(eventBook);

        String typeUrl = commandAny.getTypeUrl();

        // Check for Notification (rejection/compensation)
        if (typeUrl.endsWith("Notification")) {
            return dispatchNotification(commandAny, state);
        }

        // Execute handler
        try {
            EventBook resultBook = handler.handle(commandBook, commandAny, state, seq);
            return BusinessResponse.newBuilder()
                    .setEvents(resultBook)
                    .build();
        } catch (CommandRejectedError e) {
            throw new RouterException("Command rejected: " + e.getReason(), e);
        }
    }

    /**
     * Dispatch a notification to the rejection handler.
     */
    private BusinessResponse dispatchNotification(Any commandAny, S state) throws RouterException {
        try {
            Notification notification = commandAny.unpack(Notification.class);

            String targetDomain = "";
            String targetCommand = "";

            if (notification.hasPayload()) {
                try {
                    RejectionNotification rejection = notification.getPayload()
                            .unpack(RejectionNotification.class);
                    if (rejection.hasRejectedCommand() &&
                            rejection.getRejectedCommand().getPagesCount() > 0) {
                        CommandBook rejectedCmd = rejection.getRejectedCommand();
                        targetDomain = rejectedCmd.hasCover() ?
                                rejectedCmd.getCover().getDomain() : "";
                        targetCommand = Helpers.typeNameFromUrl(
                                rejectedCmd.getPages(0).getCommand().getTypeUrl());
                    }
                } catch (InvalidProtocolBufferException ignored) {
                    // Malformed rejection notification
                }
            }

            RejectionHandlerResponse response = handler.onRejected(
                    notification, state, targetDomain, targetCommand);

            if (response.hasEvents()) {
                return BusinessResponse.newBuilder()
                        .setEvents(response.getEvents())
                        .build();
            } else if (response.hasNotification()) {
                return BusinessResponse.newBuilder()
                        .setNotification(response.getNotification())
                        .build();
            } else {
                return BusinessResponse.newBuilder()
                        .setRevocation(RevocationResponse.newBuilder()
                                .setEmitSystemRevocation(true)
                                .setSendToDeadLetterQueue(false)
                                .setEscalate(false)
                                .setAbort(false)
                                .setReason(String.format(
                                        "Handler returned empty response for %s/%s",
                                        targetDomain, targetCommand))
                                .build())
                        .build();
            }
        } catch (InvalidProtocolBufferException e) {
            throw new RouterException("Failed to decode Notification", e);
        } catch (CommandRejectedError e) {
            throw new RouterException("Rejection handler failed: " + e.getReason(), e);
        }
    }

    /**
     * Exception type for router errors.
     */
    public static class RouterException extends Exception {
        public RouterException(String message) {
            super(message);
        }

        public RouterException(String message, Throwable cause) {
            super(message, cause);
        }
    }
}

File: client/csharp/Angzarr.Client/Router/CommandHandlerRouter.cs
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;

namespace Angzarr.Client.Router;

/// <summary>
/// Router for command handler components (commands -> events, single domain).
///
/// Domain is set at construction time. No Domain() method exists,
/// enforcing single-domain constraint at compile time.
///
/// Example:
/// <code>
/// var router = new CommandHandlerRouter&lt;PlayerState, PlayerHandler&gt;(
///     "player",
///     "player",
///     new PlayerHandler()
/// );
///
/// // Get subscriptions for framework registration
/// var subs = router.Subscriptions();
///
/// // Dispatch commands
/// var response = router.Dispatch(contextualCommand);
/// </code>
/// </summary>
/// <typeparam name="TState">The state type for this command handler.</typeparam>
/// <typeparam name="THandler">The handler implementation type.</typeparam>
public class CommandHandlerRouter<TState, THandler>
    where TState : new()
    where THandler : ICommandHandlerDomainHandler<TState>
{
    private readonly string _name;
    private readonly string _domain;
    private readonly THandler _handler;

    /// <summary>
    /// Create a new command handler router.
    /// Command handlers receive commands and emit events. Single domain enforced at construction.
    /// </summary>
    /// <param name="name">Component name.</param>
    /// <param name="domain">Domain this command handler belongs to.</param>
    /// <param name="handler">Handler implementation.</param>
    public CommandHandlerRouter(string name, string domain, THandler handler)
    {
        _name = name;
        _domain = domain;
        _handler = handler;
    }

    /// <summary>
    /// Get the router name.
    /// </summary>
    public string Name => _name;

    /// <summary>
    /// Get the domain.
    /// </summary>
    public string Domain => _domain;

    /// <summary>
    /// Get command types from the handler.
    /// </summary>
    public IReadOnlyList<string> CommandTypes() => _handler.CommandTypes();

    /// <summary>
    /// Get subscriptions for this command handler.
    /// Returns list of (domain, command types) tuples.
    /// </summary>
    public IReadOnlyList<(string Domain, IReadOnlyList<string> Types)> Subscriptions()
    {
        return new[] { (_domain, _handler.CommandTypes()) };
    }

    /// <summary>
    /// Rebuild state from events using the handler's state router.
    /// </summary>
    public TState RebuildState(Angzarr.EventBook? events)
    {
        return _handler.Rebuild(events);
    }

    /// <summary>
    /// Dispatch a contextual command to the handler.
    /// </summary>
    /// <param name="cmd">The contextual command containing command book and prior events.</param>
    /// <returns>Business response with events or rejection.</returns>
    public Angzarr.BusinessResponse Dispatch(Angzarr.ContextualCommand cmd)
    {
        var commandBook = cmd.Command;
        if (commandBook == null)
            throw new InvalidArgumentError("Missing command book");

        if (commandBook.Pages.Count == 0)
            throw new InvalidArgumentError("Missing command page");

        var commandPage = commandBook.Pages[0];
        var commandAny = commandPage.Command;
        if (commandAny == null)
            throw new InvalidArgumentError("Missing command");

        var eventBook = cmd.Events;

        // Rebuild state
        var state = _handler.Rebuild(eventBook);
        var seq = Helpers.NextSequence(eventBook);

        var typeUrl = commandAny.TypeUrl;

        // Check for Notification (rejection/compensation)
        if (typeUrl.EndsWith("Notification"))
        {
            return DispatchNotification(commandAny, state);
        }

        // Execute handler
        var resultBook = _handler.Handle(commandBook, commandAny, state, (int)seq);

        return new Angzarr.BusinessResponse { Events = resultBook };
    }

    private Angzarr.BusinessResponse DispatchNotification(Any commandAny, TState state)
    {
        var notification = commandAny.Unpack<Angzarr.Notification>();

        var rejection = notification.Payload?.Unpack<Angzarr.RejectionNotification>();
        var domain = "";
        var commandSuffix = "";

        if (rejection?.RejectedCommand?.Pages.Count > 0)
        {
            var rejectedCmd = rejection.RejectedCommand;
            domain = rejectedCmd.Cover?.Domain ?? "";
            var cmdTypeUrl = rejectedCmd.Pages[0].Command?.TypeUrl ?? "";
            commandSuffix = Helpers.TypeNameFromUrl(cmdTypeUrl);
        }

        var response = _handler.OnRejected(notification, state, domain, commandSuffix);

        if (response.Events != null)
        {
            return new Angzarr.BusinessResponse { Events = response.Events };
        }

        if (response.Notification != null)
        {
            return new Angzarr.BusinessResponse { Notification = response.Notification };
        }

        return new Angzarr.BusinessResponse
        {
            Revocation = new Angzarr.RevocationResponse
            {
                EmitSystemRevocation = true,
                SendToDeadLetterQueue = false,
                Escalate = false,
                Abort = false,
                Reason = $"Handler returned empty response for {domain}/{commandSuffix}",
            },
        };
    }
}
